### PR TITLE
[exporter/awsprometheusremotewrite] Add an option to configure STS endpoint

### PR DESCRIPTION
**Pull request by lightcode**
**Description:**
The `awsprometheusremotewriteexporter` currently use the global STS endpoint (`sts.amazonaws.com`) when a role ARN is specified. This is causing some issues when the environment is not connected to the Internet and it would be better to allow the user to choose the regional endpoint (for example `us-east-2.amazonaws.com`) for AWS PrivateLink to work.

**Testing:**
I have built the `aws-otel-collector` with my patch in order to validate the change. It works on an isolated VPC and a VPC endpoint to the regional STS. Here is the configuration used:

```yaml
  awsprometheusremotewrite:
    endpoint: ${endpoint}api/v1/remote_write
    aws_auth:
      region: ${region}
      service: aps
      role_arn: ${role_arn}
      sts_endpoint: regional
```

### CODE REVIEW

**Review by anuraaga — COMMENTED** [review]
FWIU AWS SDK uses regional STS in most regions but not older ones like us-east-1 by default. I suspect the collector should just follow the AWS defaults, and if those are bad defaults the issue could be raised to the AWS SDK. Is there a specific reason for the OTel collector to diverge from the AWS defaults?

**Comment by lightcode** [comment]
@anuraaga I found [this comment](https://github.com/aws/aws-sdk-go/issues/4112#issuecomment-941292922) explaining the behavior of the AWS SDK. I've dug into the code and found that the region `eu-central-1` is a [legacy one among others](https://github.com/aws/aws-sdk-go/blob/v1.42.22/aws/endpoints/legacy_regions.go#L10). [This function](https://github.com/aws/aws-sdk-go/blob/v1.42.22/aws/endpoints/v3model.go#L293) seems to explain why I'm getting the global STS endpoint for my region.

Why should we override this parameter? For me, if we want the OTel collector to work in air-gapped environment, we need to use the regional endpoint in order to read the STS endpoint through a VPC endpoint.

**Comment by anuraaga** [comment]
I suspect if it was a 100% seamless experience to switch the default for all regions, the AWS SDK would have done it. So I'm not very comfortable doing it here. Instead it seems appropriate for us to expose the same config the SDK does in our yaml, an sts_endpoint field. Does that make sense?

**Comment by lightcode** [comment]
I understand, it would be better to allow the override of `STSRegionalEndpoint` using an option in the config file. I will work on that in the next few days.

File: exporter/awsprometheusremotewriteexporter/auth.go [hunk 1]
```diff
@@ -104,8 +105,16 @@ func newSigningRoundTripper(cfg *Config, next http.RoundTripper, runtimeInfo str
 }
 
 func getCredsFromConfig(auth AuthConfig) (*credentials.Credentials, error) {
+	config := aws.Config{Region: aws.String(auth.Region)}
+	if auth.STSEndpoint != "" {
```
Comment (anuraaga) [inline exporter/awsprometheusremotewriteexporter/auth.go:109]: I suppose there isn't a good way to write a unit test for this logic?

Comment (Aneurysm9) [reply exporter/awsprometheusremotewriteexporter/auth.go:109]: The creation of the `aws.Config` could be extracted into a separate function and a test written to ensure that the `config.STSRegionalEndpoint` value is set appropriately.

Comment (lightcode) [reply exporter/awsprometheusremotewriteexporter/auth.go:109]: I've added a new unit test.

File: exporter/awsprometheusremotewriteexporter/config.go [hunk 2]
```diff
@@ -37,4 +37,7 @@ type AuthConfig struct {
 
 	// Amazon Resource Name (ARN) of a role to assume. Optional.
 	RoleArn string `mapstructure:"role_arn"`
+
+	// Select which endpoint to use for STS between `legacy` or `regional`. Optional.
+	STSEndpoint string `mapstructure:"sts_endpoint"`
 }
```
[the remark this answers is not in the document]
Comment (lightcode) [reply exporter/awsprometheusremotewriteexporter/config.go:45]: Done.

**Review by anuraaga — APPROVED** [review]
Thanks @lightcode - just need one test of the config validation similar to this one but otherwise looks good

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/observer/k8sobserver/config_test.go#L75

**Comment by lightcode** [comment]
> Thanks @lightcode - just need one test of the config validation similar to this one but otherwise looks good

@anuraaga I've updated my PR with the test.

**Comment by anuraaga** [comment]
Thanks @lightcode - also you'll need to sign the CLA through the link in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6713#issuecomment-991579326

**Comment by lightcode** [comment]
@anuraaga Yes, I'm working on it.

Do you know why this test fails: https://github.com/open-telemetry/opentelemetry-collector-contrib/runs/4918518101?check_suite_focus=true#step:8:62

Of course, it works on my laptop...

**Comment by anuraaga** [comment]
@lightcode Oh looks like a missing import statement

**Comment by lightcode** [comment]
@anuraaga it wouldn't have worked on my laptop then. Also, I've copy and paste the previous check and I can't find any issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/18c29df94a6e347285f7983441ca01ddee1dd717/exporter/awsprometheusremotewriteexporter/config_test.go

**Comment by Aneurysm9** [comment]
No idea what's up here.  I've re-triggered the jobs and we'll see if they feel like playing nice today.  Weirdly, the error is reported as being on the wrong line.  That is line 107 and not 108.  Perhaps it was confused by a forced push and did not re-trigger the jobs?

**Review by codeboten — COMMENTED** [review]
The tests still appear to be failing with some missing imports

```
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsprometheusremotewriteexporter [build failed]
Error: ./config_test.go:108:14: undefined: configtest
Error: ./config_test.go:108:47: undefined: path
```

**Comment by lightcode** [comment]
> The tests still appear to be failing with some missing imports
> 
> ```
> FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsprometheusremotewriteexporter [build failed]
> Error: ./config_test.go:108:14: undefined: configtest
> Error: ./config_test.go:108:47: undefined: path
> ```

Unfortunately, I can't reproduce it locally:

```
[nix-shell:~/git/github/opentelemetry-collector-contrib]$ make -C ./exporter/awsprometheusremotewriteexporter lint
make: Entering directory '/home/matthieu/git/github/opentelemetry-collector-contrib/exporter/awsprometheusremotewriteexporter'
golangci-lint run --allow-parallel-runners
make: Leaving directory '/home/matthieu/git/github/opentelemetry-collector-contrib/exporter/awsprometheusremotewriteexporter'
```

This part of the code is failing:

```go
func TestInvalidAuth(t *testing.T) {
    factories, err := componenttest.NopFactories()
    assert.NoError(t, err)
    
    factory := NewFactory()
    factories.Exporters[typeStr] = factory
    cfg, err := configtest.LoadConfigAndValidate(path.Join(".", "testdata", "invalid_sts_endpoint.yaml"), factories)
```

This is only a copy/paste of what it was done before, in the same file:

```
func TestLoadConfig(t *testing.T) {
    factories, err := componenttest.NopFactories()
    assert.NoError(t, err)
    
    factory := NewFactory()
    factories.Exporters[typeStr] = factory
    cfg, err := configtest.LoadConfigAndValidate(path.Join(".", "testdata", "config.yaml"), factories)
```

**Pull request closed without merging**